<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

# Exercise 1

In [7]:
res = []
for number in range(10):
    if number%2 == 0:
        continue
    res.append(number**2)
print(res)

[1, 9, 25, 49, 81]


# Exercise 2

In [8]:
[number**2 for number in range(0,10) if number%2 != 0]

[1, 9, 25, 49, 81]

# Exercise 3

In [9]:
%%timeit
res = []
for number in range(10):
    if number%2 == 0:
        continue
    res.append(number**2)

1.7 µs ± 164 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [8]:
%%timeit
[number**2 for number in range(0,10) if number%2 != 0]

1.69 µs ± 261 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


1.7 microseconds for the loop and 1.69 microseconds for the list comprehension method. Thefore list comprehension is faster

# Exercise 4

In [13]:
initial = 100
years = 0

while initial > 50:
    initial = initial*0.95
    years +=1

print(f"it will take {years} years")

it will take 14 years


# Exercise 5

In [3]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

In [6]:
for idx in range(len(co2_data)-1):
    y1_data = co2_data[idx]
    y2_data = co2_data[idx+1]
    conc_change = y2_data[1] - y1_data[1]
    if conc_change>3:
        string = f'{y1_data[0]} to {y2_data[0]} : {conc_change:.2f}ppm'
        print(string)

2015 to 2016 : 3.41ppm
